gtk/details.cc

```diff
@@ -1455,323 +1281,253 @@ static void refreshPeerRow(GtkListStore* store, GtkTreeIter* iter, tr_peer_stat
+            auto const key = gtr_sprintf("%d.%s", tr_torrentId(tor), s->addr);
 
-            g_snprintf(key, sizeof(key), "%d.%s", tr_torrentId(tor), s->addr);
-
-            if (g_hash_table_lookup(hash, key) == nullptr)
+            if (hash.find(key) == hash.end())
             {
-                GtkTreePath* p;
-                gtk_list_store_append(store, &iter);
-                initPeerRow(store, &iter, key, tr_torrentName(tor), s);
-                p = gtk_tree_model_get_path(model, &iter);
-                g_hash_table_insert(hash, g_strdup(key), gtk_tree_row_reference_new(model, p));
-                gtk_tree_path_free(p);
+                auto const iter = store->append();
+                initPeerRow(iter, key, tr_torrentName(tor), s);
+                hash.emplace(key, Gtk::TreeRowReference(store, store->get_path(iter)));
             }
         }
     }
 
     /* step 4: update the peers */
-    for (int i = 0; i < n; ++i)
+    for (size_t i = 0; i < torrents.size(); ++i)
     {
-        tr_torrent const* tor = torrents[i];
+        auto const* tor = torrents.at(i);
 
         for (int j = 0; j < peerCount[i]; ++j)
         {
-            char key[128];
-            GtkTreePath* p;
-            tr_peer_stat const* s = &peers[i][j];
-
-            g_snprintf(key, sizeof(key), "%d.%s", tr_torrentId(tor), s->addr);
-            auto* ref = static_cast<GtkTreeRowReference*>(g_hash_table_lookup(hash, key));
-            p = gtk_tree_row_reference_get_path(ref);
-            gtk_tree_model_get_iter(model, &iter, p);
-            refreshPeerRow(store, &iter, s);
-            gtk_tree_path_free(p);
+            auto const* s = &peers.at(i)[j];
+            auto const key = gtr_sprintf("%d.%s", tr_torrentId(tor), s->addr);
+            refreshPeerRow(store->get_iter(hash.at(key).get_path()), s);
         }
     }
 
     /* step 5: remove peers that have disappeared */
-    if (gtk_tree_model_iter_nth_child(model, &iter, nullptr, 0))
+    if (auto iter = store->children().begin(); iter)
     {
-        gboolean more = TRUE;
-
-        while (more)
+        while (iter)
         {
-            gboolean b;
-            gtk_tree_model_get(model, &iter, PEER_COL_WAS_UPDATED, &b, -1);
-
-            if (b)
+            if (iter->get_value(peer_cols.was_updated))
             {
-                more = gtk_tree_model_iter_next(model, &iter);
+                ++iter;
             }
             else
             {
-                char* key;
-                gtk_tree_model_get(model, &iter, PEER_COL_KEY, &key, -1);
-                g_hash_table_remove(hash, key);
-                more = gtk_list_store_remove(store, &iter);
-                g_free(key);
+                auto const key = iter->get_value(peer_cols.key);
+                hash.erase(key);
+                iter = store->erase(iter);
             }
         }
     }
 
     /* step 6: cleanup */
-    for (int i = 0; i < n; ++i)
+    for (size_t i = 0; i < peers.size(); ++i)
```

Comment:
range loop

Reply:
That would require combining `peers` and `peerCount` vectors, which is a bigger change I didn't want to do at this time.